File: gandhara-site/scripts/import-from-wp-sql.py
#!/usr/bin/env python3
"""Import WordPress SQL dump into Astro JSON data files."""

from __future__ import annotations

import json
import re
from pathlib import Path

ROOT = Path(__file__).resolve().parent.parent
SQL_PATH = ROOT.parent / "Gand_" / "dbppgfjxviqljp.sql"
DATA_DIR = ROOT / "src" / "data"
UPLOADS_DIR = ROOT / "public" / "uploads"

EXCLUDED_MEDIA = ["SFILATA-SMORZO-HD-1080p.mov"]

EXCLUDED_VIDEO_PLACEHOLDER = (
    "<!-- VIDEO ESCLUSO (file pesante, reinserire in public/uploads): "
    "/uploads/2021/11/SFILATA-SMORZO-HD-1080p.mov -->"
)

TEXT_ONLY_SLUGS = {"presentazione"}

PRESENTAZIONE_PLACEHOLDER = (
    "<!-- PRESENTAZIONE: galleria immagini esclusa (file pesanti). "
    "Reinserire le foto da /uploads/2022/06/ (nikon-fm2-x-Gandhara0001, ecc.) -->"
)

SLUG_MAP: dict[str, dict[str, dict]] = {
    "pharmakon": {
        "pharmakon-ep2": {"post_name": "pharmakon-ep-2", "post_type": "pharmakon"},
        "pharmakon-ep1": {"post_name": "articolo-1-pharmakon", "post_type": "pharmakon"},
        "presentazione": {"post_name": "presentazione", "post_type": "pharmakon"},
    },
    "stampa": {
        "pharmakon-ep1": {"post_name": "articolo-1-stampa", "post_type": "stampa"},
        "altro": {"post_name": "altro", "post_type": "stampa"},
    },
    "collab": {
        "una-vetrina": {"post_name": "prova-collab", "post_type": "collab"},
        "villa-medici": {"post_name": "villa-medici", "post_type": "collab"},
        "haus-of-dreamers": {"post_name": "haus-of-dreamers", "post_type": "collab"},
        "wegil": {"post_name": "wegil", "post_type": "collab"},
    },
    "mostre": {
        "smorzo": {"post_name": "smorzo", "post_type": "post", "post_id": 744},
        "altri-luoghi": {"post_name": "altri-luoghi", "post_type": "post", "post_id": 716},
        "habicura": {"post_name": "habicura-giardini-verano", "post_type": "post", "post_id": 763},
        "extrart": {"post_name": "extrart", "post_type": "post"},
        "instant-paper": {"post_name": "gandhara-slice-instant-paper", "post_type": "post"},
        "riflettiti": {"post_name": "riflettiti", "post_type": "post"},
        "naked-nature": {"post_name": "naked-nature", "post_type": "post"},
    },
    "cataloghi": {
        "smorzo": {"post_name": "smorzo", "post_type": "cataloghi", "post_id": 651},
        "altri-luoghi": {"post_name": "altri-luoghi", "post_type": "cataloghi"},
        "abitare-linabitabile": {"post_name": "abitare-linabitabile", "post_type": "cataloghi", "post_id": 742},
    },
}


def parse_sql_row(row: str) -> list[str | None]:
    fields: list[str | None] = []
    i = 1 if row.startswith("(") else 0
    n = len(row)

    while i < n and len(fields) < 22:
        while i < n and row[i] in " \t\r\n":
            i += 1
        if i >= n:
            break
        if row[i : i + 4] == "NULL":
            fields.append(None)
            i += 4
            continue
        if row[i] != "'":
            j = i
            while j < n and row[j] in "0123456789-":
                j += 1
            fields.append(row[i:j])
            i = j
            while i < n and row[i] in ", \t\r\n":
                i += 1
            continue
        i += 1
        chars: list[str] = []
        while i < n:
            c = row[i]
            if c == "\\" and i + 1 < n:
                nxt = row[i + 1]
                if nxt == "'":
                    chars.append("'")
                    i += 2
                    continue
                if nxt == "n":
                    chars.append("\n")
                    i += 2
                    continue
                if nxt == "r":
                    chars.append("\r")
                    i += 2
                    continue
                if nxt == "t":
                    chars.append("\t")
                    i += 2
                    continue
                if nxt == "\\":
                    chars.append("\\")
                    i += 2
                    continue
                chars.append(nxt)
                i += 2
                continue
            if c == "'":
                if i + 1 < n and row[i + 1] == "'":
                    chars.append("'")
                    i += 2
                    continue
                i += 1
                break
            chars.append(c)
            i += 1
        fields.append("".join(chars))
        while i < n and row[i] in ", \t\r\n":
            i += 1
    return fields


def load_posts(sql: str) -> tuple[dict, dict]:
    posts: dict = {}
    attachments: dict[int, str] = {}
    header = "INSERT INTO `kzc_posts`"
    pos = 0

    while True:
        start = sql.find(header, pos)
        if start < 0:
            break
        values_idx = sql.find("VALUES", start)
        next_insert = sql.find("\nINSERT INTO `", values_idx + 6)
        block = sql[values_idx + 6 : next_insert if next_insert > 0 else len(sql)]
        pos = start + len(header)

        # Each post row is a single line in this dump
        for line in block.split("\n"):
            slice_ = line.strip().rstrip(",").rstrip(";")
            if not slice_.startswith("("):
                continue
            fields = parse_sql_row(slice_)
            if len(fields) < 21 or not fields[0] or not str(fields[0]).isdigit():
                continue
            post = {
                "ID": int(fields[0]),
                "post_content": fields[4] or "",
                "post_title": fields[5] or "",
                "post_status": fields[7] or "",
                "post_name": fields[11] or "",
                "post_type": fields[20] or "",
                "guid": fields[18] or "",
            }
            if post["post_type"] == "attachment":
                match = re.search(r"uploads/(.+)$", post["guid"], re.I)
                if match:
                    attachments[post["ID"]] = match.group(1)
            else:
                posts[post["ID"]] = post
                posts[f"{post['post_type']}:{post['post_name']}"] = post

    return posts, attachments


def load_postmeta(sql: str) -> dict[int, dict[str, str]]:
    meta: dict[int, dict[str, str]] = {}
    for m in re.finditer(
        r"\((\d+),\s*(\d+),\s*'((?:[^'\\]|\\.|'')*)',\s*'((?:[^'\\]|\\.|'')*)'\)",
        sql,
    ):
        post_id = int(m.group(2))
        key = m.group(3).replace("\\'", "'").replace("''", "'")
        val = m.group(4).replace("\\'", "'").replace("''", "'")
        meta.setdefault(post_id, {})[key] = val
    return meta


def list_upload_files(directory: Path, base: str = "/uploads") -> set[str]:
    out: set[str] = set()
    if not directory.exists():
        return out
    for p in directory.rglob("*"):
        if p.is_file():
            rel = f"{base}/{p.relative_to(directory).as_posix()}"
            out.add(rel)
    return out


def resolve_upload_url(url_path: str, existing: set[str]) -> str:
    if not url_path:
        return url_path
    p = re.sub(r"^https?://[^/]+", "", url_path, flags=re.I)
    if "wp-content/uploads/" in p:
        p = "/uploads/" + p.split("wp-content/uploads/")[1]
    if not p.startswith("/uploads/"):
        idx = p.find("/uploads/")
        if idx >= 0:
            p = p[idx:]
    if p in existing:
        return p
    base = re.sub(r"-\d+x\d+(\.[a-z0-9]+)$", r"\1", p, flags=re.I)
    if base in existing:
        return base
    return p


def rewrite_urls(html: str, existing: set[str]) -> str:
    def repl(m: re.Match) -> str:
        path = m.group(1)
        return resolve_upload_url("/uploads" + path.replace(" ", "%20"), existing)

    html = re.sub(
        r"(?:https?:)?//[^/]+/wp-content/uploads(/[^\s\"'<>]+)",
        repl,
        html,
        flags=re.I,
    )
    html = re.sub(
        r"/wp-content/uploads(/[^\s\"'<>]+)",
        repl,
        html,
        flags=re.I,
    )

    def src_repl(m: re.Match) -> str:
        return f'src="{resolve_upload_url(m.group(1), existing)}"'

    return re.sub(r'src="([^"]+)"', src_repl, html)


def strip_excluded_media(html: str) -> str:
    out = html
    for pattern in EXCLUDED_MEDIA:
        esc = re.escape(pattern)
        out = re.sub(
            rf"<!-- wp:video[^]*?{esc}[^]*?<!-- /wp:video -->",
            EXCLUDED_VIDEO_PLACEHOLDER,
            out,
            flags=re.I | re.DOTALL,
        )
        out = re.sub(
            rf"<figure class=\"wp-block-video\">[^]*?{esc}[^]*?</figure>",
            EXCLUDED_VIDEO_PLACEHOLDER,
            out,
            flags=re.I | re.DOTALL,
        )
        out = re.sub(
            rf"<video[^>]*src=\"[^\"]*{esc}[^\"]*\"[^>]*>.*?</video>",
            EXCLUDED_VIDEO_PLACEHOLDER,
            out,
            flags=re.I | re.DOTALL,
        )
        out = re.sub(
            rf"<figure class=\"wp-block-video\">[^]*?{esc}[^]*?</figure>",
            EXCLUDED_VIDEO_PLACEHOLDER,
            out,
            flags=re.I | re.DOTALL,
        )
    return out


def gutenberg_to_html(content: str, *, text_only: bool = False) -> str:
    if not content:
        return ""
    html = re.sub(r"<!-- /wp:[^>]+ -->", "", content)
    html = re.sub(r"<!-- wp:[^>]+ -->", "", html)
    html = re.sub(r'<p class="wp-block-pdfemb[^"]*"></p>', "", html)
    html = re.sub(
        r'<figure class="wp-block-image"><img alt=""\s*/?></figure>',
        "",
        html,
        flags=re.I,
    )

    if text_only:
        parts: list[str] = []
        for m in re.finditer(
            r"<p[^>]*>([\s\S]*?)</p>|<h([1-6])[^>]*>([\s\S]*?)</h\2>",
            html,
            re.I,
        ):
            if m.group(1):
                inner = m.group(1)
                if inner.strip() and "wp-image" not in inner:
                    parts.append(f"<p>{inner}</p>")
            elif m.group(3):
                parts.append(f"<h4>{m.group(3)}</h4>")
        return PRESENTAZIONE_PLACEHOLDER + "\n" + "\n".join(parts)

    def gallery_repl(m: re.Match) -> str:
        imgs = re.findall(r"<img[^>]+>", m.group(1), re.I)
        return "\n".join(f'<figure class="wp-block-image">{img}</figure>' for img in imgs)

    html = re.sub(
        r'<ul class="blocks-gallery-grid">([\s\S]*?)</ul>',
        gallery_repl,
        html,
        flags=re.I,
    )
    html = re.sub(r'<div class="wp-block-column"[^>]*>\s*</div>', "", html, flags=re.I)
    html = re.sub(r"<p>\s*</p>", "", html, flags=re.I)
    return html.strip()


def find_post(posts: dict, cfg: dict):
    post_id = cfg.get("post_id")
    if post_id and post_id in posts:
        return posts[post_id]
    key = f"{cfg['post_type']}:{cfg['post_name']}"
    if key in posts:
        return posts[key]
    for p in posts.values():
        if (
            isinstance(p, dict)
            and p.get("post_name") == cfg["post_name"]
            and p.get("post_type") == cfg["post_type"]
            and p.get("post_status") == "publish"
        ):
            return p
    return None


def get_cover_image(
    post_id: int,
    meta: dict,
    attachments: dict,
    content: str,
    existing: set[str],
) -> str:
    m = meta.get(post_id, {})
    att_id = m.get("immagine_copertina") or m.get("_thumbnail_id")
    if att_id:
        aid = int(att_id)
        if aid in attachments:
            return resolve_upload_url(f"/uploads/{attachments[aid]}", existing)
    img = re.search(r'src="(/uploads/[^"]+)"', content)
    if img:
        return img.group(1)
    wp_img = re.search(r"uploads/([^\"'\s]+)", content)
    if wp_img:
        return resolve_upload_url(f"/uploads/{wp_img.group(1)}", existing)
    return ""


def import_section(
    section: str,
    posts: dict,
    meta: dict,
    attachments: dict,
    existing: set[str],
) -> None:
    mapping = SLUG_MAP[section]
    json_path = DATA_DIR / f"{section}.json"
    items = json.loads(json_path.read_text(encoding="utf-8"))
    results = []

    for item in items:
        cfg = mapping.get(item["slug"])
        if not cfg:
            print(f"  [{section}] no mapping: {item['slug']}")
            results.append(item)
            continue
        post = find_post(posts, cfg)
        if not post:
            print(f"  [{section}] not found: {item['slug']}")
            results.append(item)
            continue

        text_only = item["slug"] in TEXT_ONLY_SLUGS
        raw = strip_excluded_media(post["post_content"])
        content = gutenberg_to_html(raw, text_only=text_only)
        content = strip_excluded_media(content)
        content = rewrite_urls(content, existing)
        cover = (
            ""
            if text_only
            else get_cover_image(post["ID"], meta, attachments, content, existing)
        )
        results.append(
            {
                "slug": item["slug"],
                "title": post["post_title"] or item["title"],
                "coverImage": cover,
                "content": content,
            }
        )
        print(f"  ✓ {section}/{item['slug']} ({len(content)} chars)")

    json_path.write_text(
        json.dumps(results, ensure_ascii=False, indent=2) + "\n",
        encoding="utf-8",
    )


def import_about(posts: dict, existing: set[str]) -> None:
    post = find_post(posts, {"post_name": "home", "post_type": "page"})
    if not post:
        return
    content = rewrite_urls(gutenberg_to_html(post["post_content"]), existing)
    (DATA_DIR / "about.json").write_text(
        json.dumps({"content": content}, ensure_ascii=False, indent=2) + "\n",
        encoding="utf-8",
    )
    print(f"  ✓ about.json ({len(content)} chars)")


def main() -> None:
    print("Import from WordPress SQL\n")
    if not SQL_PATH.exists():
        raise SystemExit(f"SQL not found: {SQL_PATH}")

    sql = SQL_PATH.read_text(encoding="utf-8", errors="replace")
    posts, attachments = load_posts(sql)
    meta = load_postmeta(sql)
    existing = list_upload_files(UPLOADS_DIR)

    post_count = len({k for k, v in posts.items() if isinstance(k, int)})
    print(f"Posts: {post_count}, attachments: {len(attachments)}")
    print(f"Upload files on disk: {len(existing)}\n")

    for section in ["pharmakon", "stampa", "collab", "mostre", "cataloghi"]:
        print(f"[{section}]")
        import_section(section, posts, meta, attachments, existing)

    print("\n[about]")
    import_about(posts, existing)
    print("\nDone.")


if __name__ == "__main__":
    main()

File: gandhara-site/scripts/import-from-wp-sql.mjs
#!/usr/bin/env node
/**
 * Import WordPress content from SQL dump into Astro JSON data files.
 * Run: node scripts/import-from-wp-sql.mjs
 */

import fs from 'fs';
import path from 'path';
import { fileURLToPath } from 'url';

const __dirname = path.dirname(fileURLToPath(import.meta.url));
const ROOT = path.join(__dirname, '..');
const SQL_PATH = path.join(ROOT, '..', 'Gand_', 'dbppgfjxviqljp.sql');
const DATA_DIR = path.join(ROOT, 'src', 'data');
const UPLOADS_DIR = path.join(ROOT, 'public', 'uploads');

/** Media esclusi temporaneamente (file pesanti rimossi dal repo). */
const EXCLUDED_MEDIA = [
  'SFILATA-SMORZO-HD-1080p.mov',
];

const EXCLUDED_VIDEO_PLACEHOLDER =
  '<!-- VIDEO ESCLUSO (file pesante, reinserire in public/uploads): /uploads/2021/11/SFILATA-SMORZO-HD-1080p.mov -->';

/** Voci con contenuto ridotto: solo testo, galleria esclusa. */
const TEXT_ONLY_SLUGS = new Set(['presentazione']);

const PRESENTAZIONE_PLACEHOLDER =
  '<!-- PRESENTAZIONE: galleria immagini esclusa (file pesanti). Reinserire le foto da /uploads/2022/06/ (nikon-fm2-x-Gandhara0001, ecc.) -->';

/** slug nel JSON Astro → { wp post_name, wp post_type, opzionale post_id alternativo } */
const SLUG_MAP = {
  pharmakon: {
    'pharmakon-ep2': { post_name: 'pharmakon-ep-2', post_type: 'pharmakon' },
    'pharmakon-ep1': { post_name: 'articolo-1-pharmakon', post_type: 'pharmakon' },
    presentazione: { post_name: 'presentazione', post_type: 'pharmakon' },
  },
  stampa: {
    'pharmakon-ep1': { post_name: 'articolo-1-stampa', post_type: 'stampa' },
    altro: { post_name: 'altro', post_type: 'stampa' },
  },
  collab: {
    'una-vetrina': { post_name: 'prova-collab', post_type: 'collab' },
    'villa-medici': { post_name: 'villa-medici', post_type: 'collab' },
    'haus-of-dreamers': { post_name: 'haus-of-dreamers', post_type: 'collab' },
    wegil: { post_name: 'wegil', post_type: 'collab' },
  },
  mostre: {
    smorzo: { post_name: 'smorzo', post_type: 'post', post_id: 744 },
    'altri-luoghi': { post_name: 'altri-luoghi', post_type: 'post', post_id: 716 },
    habicura: { post_name: 'habicura-giardini-verano', post_type: 'post', post_id: 763 },
    extrart: { post_name: 'extrart', post_type: 'post' },
    'instant-paper': { post_name: 'gandhara-slice-instant-paper', post_type: 'post' },
    riflettiti: { post_name: 'riflettiti', post_type: 'post' },
    'naked-nature': { post_name: 'naked-nature', post_type: 'post' },
  },
  cataloghi: {
    smorzo: { post_name: 'smorzo', post_type: 'cataloghi', post_id: 651 },
    'altri-luoghi': { post_name: 'altri-luoghi', post_type: 'cataloghi' },
    'abitare-linabitabile': { post_name: 'inabitabile', post_type: 'cataloghi' },
  },
};

// --- SQL parsing ---

function parseSqlStringFields(row) {
  const fields = [];
  let i = 0;
  if (row[0] === '(') i = 1;
  while (i < row.length && fields.length < 22) {
    if (row[i] === 'N' && row.slice(i, i + 4) === 'NULL') {
      fields.push(null);
      i += 4;
      if (row[i] === ',') i++;
      continue;
    }
    if (row[i] !== "'") {
      // unquoted number
      let j = i;
      while (j < row.length && /[0-9-]/.test(row[j])) j++;
      fields.push(row.slice(i, j));
      i = j;
      if (row[i] === ',') i++;
      continue;
    }
    i++;
    let s = '';
    while (i < row.length) {
      const c = row[i];
      if (c === '\\' && i + 1 < row.length) {
        const next = row[i + 1];
        if (next === "'") {
          s += "'";
          i += 2;
          continue;
        }
        if (next === 'n') {
          s += '\n';
          i += 2;
          continue;
        }
        if (next === 'r') {
          s += '\r';
          i += 2;
          continue;
        }
        if (next === 't') {
          s += '\t';
          i += 2;
          continue;
        }
        if (next === '\\') {
          s += '\\';
          i += 2;
          continue;
        }
        s += next;
        i += 2;
        continue;
      }
      if (c === "'") {
        if (row[i + 1] === "'") {
          s += "'";
          i += 2;
          continue;
        }
        i++;
        break;
      }
      s += c;
      i++;
    }
    fields.push(s);
    if (row[i] === ',') i++;
  }
  return fields;
}

function loadPosts(sql) {
  const insertHeader = 'INSERT INTO `kzc_posts`';
  const blocks = [];
  let pos = 0;
  while (true) {
    const start = sql.indexOf(insertHeader, pos);
    if (start < 0) break;
    const valuesIdx = sql.indexOf('VALUES', start);
    const nextInsert = sql.indexOf('\nINSERT INTO `', valuesIdx + 6);
    const block = sql.slice(valuesIdx + 6, nextInsert > 0 ? nextInsert : sql.length);
    blocks.push(block);
    pos = start + insertHeader.length;
  }

  const posts = new Map();
  const attachments = new Map();

  for (const block of blocks) {
  const rowRe = /^\(\d+,/gm;
  let m;
  const starts = [];
  while ((m = rowRe.exec(block)) !== null) starts.push(m.index);

  for (let i = 0; i < starts.length; i++) {
    const slice = block.slice(starts[i], starts[i + 1] ?? block.length).replace(/\),\s*$/, '').replace(/\);?\s*$/, '');
    const fields = parseSqlStringFields(slice);
    if (fields.length < 21) continue;
    const post = {
      ID: parseInt(fields[0], 10),
      post_content: fields[4] ?? '',
      post_title: fields[5] ?? '',
      post_status: fields[7] ?? '',
      post_name: fields[11] ?? '',
      post_type: fields[20] ?? '',
      guid: fields[18] ?? '',
    };
    if (post.post_type === 'attachment') {
      const guid = post.guid;
      const match = guid.match(/uploads\/(.+)$/i) || post.post_content.match(/uploads\/(.+)/i);
      if (match) attachments.set(post.ID, match[1]);
    } else {
      posts.set(post.ID, post);
      const key = `${post.post_type}:${post.post_name}`;
      posts.set(key, post);
    }
  }
  }
  return { posts, attachments };
}

function loadPostMeta(sql) {
  const meta = new Map();
  const re = /\((\d+),\s*(\d+),\s*'([^']+)',\s*'((?:[^'\\]|\\.|'')*)'\)/g;
  let m;
  while ((m = re.exec(sql)) !== null) {
    const postId = parseInt(m[2], 10);
    const key = m[3].replace(/\\'/g, "'").replace(/''/g, "'");
    let val = m[4].replace(/\\'/g, "'").replace(/''/g, "'");
    if (!meta.has(postId)) meta.set(postId, {});
    meta.get(postId)[key] = val;
  }
  return meta;
}

function listUploadFiles(dir, base = '/uploads') {
  const out = new Set();
  if (!fs.existsSync(dir)) return out;
  for (const ent of fs.readdirSync(dir, { withFileTypes: true })) {
    const p = path.join(dir, ent.name);
    if (ent.isDirectory()) {
      for (const x of listUploadFiles(p, `${base}/${ent.name}`)) out.add(x);
    } else {
      out.add(`${base}/${ent.name}`);
    }
  }
  return out;
}

function resolveUploadUrl(urlPath, existingFiles) {
  if (!urlPath) return urlPath;
  let p = urlPath.replace(/^https?:\/\/[^/]+/i, '');
  if (!p.startsWith('/uploads/')) {
    const idx = p.indexOf('/uploads/');
    if (idx >= 0) p = p.slice(idx);
    else if (p.includes('wp-content/uploads/')) {
      p = '/uploads/' + p.split('wp-content/uploads/')[1];
    }
  }
  p = p.replace(/ /g, '%20');
  if (existingFiles.has(p)) return p;
  // try without size suffix
  const base = p.replace(/-\d+x\d+(\.[a-z0-9]+)$/i, '$1');
  if (existingFiles.has(base)) return base;
  return p;
}

function rewriteUrls(html, existingFiles) {
  return html.replace(
    /(?:https?:)?\/\/(?:gandhara\.info\/wp-content\/uploads|gandhara\.info\/wp-content\/uploads|\/wp-content\/uploads)(\/[^\s"'<>]+)/gi,
    (_, rest) => resolveUploadUrl('/uploads' + rest.replace(/ /g, '%20'), existingFiles),
  ).replace(
    /src="([^"]+)"/g,
    (_, u) => `src="${resolveUploadUrl(u, existingFiles)}"`,
  );
}

function stripExcludedMedia(html) {
  let out = html;
  for (const pattern of EXCLUDED_MEDIA) {
    const re = new RegExp(
      `<!-- wp:video[^]*?${pattern.replace(/[.*+?^${}()|[\]\\]/g, '\\$&')}[^]*?<!-- /wp:video -->`,
      'gi',
    );
    out = out.replace(re, EXCLUDED_VIDEO_PLACEHOLDER);
    out = out.replace(
      new RegExp(`<figure class="wp-block-video">[^]*?${pattern.replace(/[.*+?^${}()|[\]\\]/g, '\\$&')}[^]*?</figure>`, 'gi'),
      EXCLUDED_VIDEO_PLACEHOLDER,
    );
  }
  return out;
}

function gutenbergToHtml(content, { textOnly = false } = {}) {
  if (!content) return '';

  let html = content;

  // Remove Gutenberg block comments, keep inner HTML
  html = html.replace(/<!-- \/wp:[^>]+ -->/g, '');
  html = html.replace(/<!-- wp:[^>]+ -->/g, '');

  // PDF embed → link
  html = html.replace(
    /<p class="wp-block-pdfemb[^"]*"><\/p>/g,
    '',
  );

  // Empty wp blocks
  html = html.replace(/<figure class="wp-block-image"><img alt=""\s*\/?><\/figure>/gi, '');

  if (textOnly) {
    // Keep only paragraphs, headings, links
    const parts = [];
    const pRe = /<p[^>]*>([\s\S]*?)<\/p>|<h[1-6][^>]*>([\s\S]*?)<\/h[1-6]>|<em>([\s\S]*?)<\/em>/gi;
    let m;
    while ((m = pRe.exec(html)) !== null) {
      const inner = m[1] || m[2] || m[3];
      if (inner && inner.trim() && !inner.includes('wp-image')) {
        if (m[2]) parts.push(`<h4>${inner}</h4>`);
        else parts.push(`<p>${inner}</p>`);
      }
    }
    return PRESENTAZIONE_PLACEHOLDER + '\n' + parts.join('\n');
  }

  // Simplify gallery grids to stacked figures
  html = html.replace(/<ul class="blocks-gallery-grid">([\s\S]*?)<\/ul>/gi, (_, inner) => {
    const imgs = [...inner.matchAll(/<img[^>]+>/gi)].map((x) => x[0]);
    return imgs.map((img) => `<figure class="wp-block-image">${img}</figure>`).join('\n');
  });

  // Collapse excessive empty columns/divs from theme blocks
  html = html.replace(/<div class="wp-block-column"[^>]*>\s*<\/div>/gi, '');
  html = html.replace(/<p>\s*<\/p>/gi, '');

  return html.trim();
}

function findPost(posts, { post_name, post_type, post_id }) {
  if (post_id && posts.has(post_id)) return posts.get(post_id);
  const key = `${post_type}:${post_name}`;
  if (posts.has(key)) return posts.get(key);
  for (const p of posts.values()) {
    if (typeof p === 'object' && p.post_name === post_name && p.post_type === post_type && p.post_status === 'publish') {
      return p;
    }
  }
  return null;
}

function getCoverImage(postId, meta, attachments, content, existingFiles) {
  const m = meta.get(postId) || {};
  let attId = m.immagine_copertina || m._thumbnail_id;
  if (attId && attachments.has(parseInt(attId, 10))) {
    const rel = '/uploads/' + attachments.get(parseInt(attId, 10));
    return resolveUploadUrl(rel, existingFiles);
  }
  const img = content.match(/src="(\/uploads\/[^"]+)"/);
  if (img) return img[1];
  const wpImg = content.match(/uploads\/([^"'\s]+)/);
  if (wpImg) return resolveUploadUrl('/uploads/' + wpImg[1], existingFiles);
  return '';
}

function importSection(section, posts, meta, attachments, existingFiles) {
  const map = SLUG_MAP[section];
  const jsonPath = path.join(DATA_DIR, `${section}.json`);
  const existing = JSON.parse(fs.readFileSync(jsonPath, 'utf8'));
  const results = [];

  for (const item of existing) {
    const cfg = map[item.slug];
    if (!cfg) {
      console.warn(`  [${section}] no WP mapping for slug: ${item.slug}`);
      results.push(item);
      continue;
    }
    const post = findPost(posts, cfg);
    if (!post) {
      console.warn(`  [${section}] post not found: ${item.slug}`);
      results.push({ ...item, content: item.content || '' });
      continue;
    }

    const textOnly = TEXT_ONLY_SLUGS.has(item.slug);
    let content = gutenbergToHtml(post.post_content, { textOnly });
    content = stripExcludedMedia(content);
    content = rewriteUrls(content, existingFiles);

    const coverImage = textOnly
      ? ''
      : getCoverImage(post.ID, meta, attachments, content, existingFiles);

    results.push({
      slug: item.slug,
      title: post.post_title || item.title,
      coverImage,
      content,
    });
    console.log(`  ✓ ${section}/${item.slug} (${content.length} chars)`);
  }

  fs.writeFileSync(jsonPath, JSON.stringify(results, null, 2) + '\n');
}

function importAbout(posts, existingFiles) {
  const home = findPost(posts, { post_name: 'home', post_type: 'page' });
  if (!home) return;
  let content = gutenbergToHtml(home.post_content);
  content = rewriteUrls(content, existingFiles);
  fs.writeFileSync(
    path.join(DATA_DIR, 'about.json'),
    JSON.stringify({ content }, null, 2) + '\n',
  );
  console.log(`  ✓ about.json (${content.length} chars)`);
}

// --- main ---

console.log('Import from WordPress SQL\n');
if (!fs.existsSync(SQL_PATH)) {
  console.error('SQL not found:', SQL_PATH);
  process.exit(1);
}

const sql = fs.readFileSync(SQL_PATH, 'utf8');
const { posts, attachments } = loadPosts(sql);
const meta = loadPostMeta(sql);
const existingFiles = listUploadFiles(UPLOADS_DIR);

const postCount = [...posts.values()].filter((p) => typeof p === 'object' && p.ID && p.post_status).length;
console.log(`Posts: ${postCount}, attachments: ${attachments.size}`);
console.log(`Upload files on disk: ${existingFiles.size}\n`);

for (const section of ['pharmakon', 'stampa', 'collab', 'mostre', 'cataloghi']) {
  console.log(`[${section}]`);
  importSection(section, posts, meta, attachments, existingFiles);
}

console.log('\n[about]');
importAbout(posts, existingFiles);

console.log('\nDone.');
